src/Mvc/Mvc.Core/src/Formatters/SystemTextJsonOutputFormatter.cs

```diff
@@ -59,7 +59,12 @@ public sealed override async Task WriteResponseBodyAsync(OutputFormatterWriteCon
             var writeStream = GetWriteStream(httpContext, selectedEncoding);
             try
             {
-                await JsonSerializer.WriteAsync(writeStream, context.Object, context.ObjectType, SerializerOptions);
+                // context.ObjectType reflects the declared model type when specified.
+                // For polymorphic scenarios where the user declares a return type, but returns a derived type,
+                // we want to serialize all the properties on the derived type. This keeps parity with
+                // the behavior you get when the user does not declare the return type and with Json.Net at least at the top level.
+                var objectType = context.Object?.GetType() ?? context.ObjectType;
```

Comment:
What do you think? I considered leaving it as is, but you get a different outcome depending on whether you do `ActionResult<T>` versus returning an `ObjectResult`. Doing this felt like it was moderately better. Polymorphic serialization does not work for properties, but I think there's plans for a feature there.

Reply:
I think this is the right thing to do given that JSON.net does it.